File: lintcode/Interleaving_Positive_and_Negative_Numbers.cpp
class Solution {
public:
    /**
     * @param A: An integer array.
     * @return an integer array
     */
    vector<int> rerange(vector<int> &A) {
        // write your code here
        if(A.size() <= 2) return A;
        
        int k = partition(A);
        int x = k;
        int y = A.size()-k;
        reverse(A, x, y);
        
        return A;
    }
private:
    int partition(vector<int> &A){
        int i = 0;
        int j = A.size()-1;
        while(i <= j){
            if(A[i] < 0) i++;
            else if(A[i] > 0) swap(A[i], A[j--]);
            else throw runtime_error("has 0 inside, wrong input!");
        }
        return i;
    }
    
    void reverse(vector<int> &A, int x, int y){
        int i = 0;
        int j = 0;
        
        if(x == y){
            i = 1;
            j = A.size()-2;
        }else if(x < y){
            i = 0;
            j = A.size()-2;
        }else{
            i = 1;
            j = A.size()-1;
        }
        
        while(i < j){
            swap(A[i], A[j]);
            i += 2;
            j -= 2;
        }
    }
};

File: lintcode/java/InterleavingPositiveandNegativeNumbers.java
class Solution {
    /**
     * @param A: An integer array.
     * @return: void
     */
    public void rerange(int[] A) {
        if(A.length <= 2) return;
        
        int n = A.length;
        int x = partation(A);
        int y = n - x;
        rerangeHelper(A, x, y);
    }
    
    private void rerangeHelper(int[] A, int x, int y){
        int n = A.length;
        
        int i = 0, j = 0;
        if(x < y){
            i = 0;
            j = n-2;
        }else if(x == y){
            i = 1;
            j = n-2;
        }else{
            i = 1;
            j = n-1;
        }
        
        while(i < j){
            swap(A, i, j);
            i += 2;
            j -= 2;
        }
    }
   
   private int partation(int[] A){
       int i = 0;
       int j = A.length -1;
       
       while(i <= j){
           if(A[i] < 0) i++;
           else swap(A, i, j--);
       }
       return i;
   }
   
   private void swap(int[] A, int i, int j){
       int tmp = A[i];
       A[i] = A[j];
       A[j] = tmp;
   }
}
